ledger/src/blockstore_processor.rs

```diff
@@ -976,14 +976,22 @@ fn load_frozen_forks(
                     ).and_then(|supermajority_root| {
                         if supermajority_root > *root {
                             // If there's a cluster confirmed root greater than our last
-                            // replayed root, then beccause the cluster confirmed root should
+                            // replayed root, then because the cluster confirmed root should
                             // be descended from our last root, it must exist in `all_banks`
                             let cluster_root_bank = all_banks.get(&supermajority_root).unwrap();
 
                             // cluster root must be a descendant of our root, otherwise something
                             // is drastically wrong
                             assert!(cluster_root_bank.ancestors.contains_key(root));
                             info!("blockstore processor found new cluster confirmed root: {}, observed in bank: {}", cluster_root_bank.slot(), bank.slot());
+
+                            // Ensure cluster-confirmed root is set as root in blockstore
+                            if !blockstore.is_root(*root) {
```

Comment:
context: https://github.com/solana-labs/solana/pull/17402#issuecomment-846568708

@CriesofCarrots Hmm, maybe should we use `supermajority_root` instead of `*root`?

@carllin could you double-check this?

Reply:
Also, is this fix confirmed like this?

1. start mainnet-beta node
2. kill the node after 1 min after rpc is brought up (so, repair has been fetched some slots into blockstore)
3. start the node and observed some blockstore_processor workings
4. observe gap or no gap of roots?

Reply:
Oof, yes, my dumb mistake! I originally had this down at L1010 and didn't edit the var name when I moved it.

Reply:
@ryoqun, @CriesofCarrots  I think you also need to set root for all parent slots in between the new `supermajority_root` and the previous root, similar to how its done in ReplayStage: https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L1324-L1331. This is because the supermajority root can skip a few blocks like so

`1 (old root) -> 2 -> 3 -> 4 (supermajority root detected at slot 5) -> 5`

Reply:
Ah, makes sense. Updated